dash/development/build_process.py

```diff
@@ -0,0 +1,210 @@
+    def bundles(self):
+        if not os.path.exists(self.build_folder):
+            try:
+                os.makedirs(self.build_folder)
+            except OSError:
+                logger.exception(
+                    "🚨 having issues manipulating %s", self.build_folder
+                )
+                sys.exit(1)
+
+        if self.__class__.__name__ == "DCC":
+            shutil.copyfile(
+                self._concat(self.main, "assets", "highlight.pack.js"),
+                self._concat(self.build_folder, "highlight.pack.js"),
+            )
+            logger.info("copy the customized highligh js")
+
+        versions = {
+            "version": self.version,
+            "package": self.name.replace(" ", "_").replace("-", "_"),
+        }
+        for name, subfolder, filename, target in self.deps_info:
+            version = self.deps[name]["version"]
+            versions[name.replace("-", "").replace(".", "")] = version
+
+            logger.info("copy npm dependency => %s", filename)
+            ext = "min.js" if "min" in filename.split(".") else "js"
+            target = (
+                target.format(version)
+                if target
+                else "{}@{}.{}".format(name, version, ext)
+            )
+            shutil.copyfile(
+                self._concat(self.npm_modules, name, subfolder, filename),
+                self._concat(self.build_folder, target),
+            )
+
+        logger.info("run `npm run build:webpack`")
+        os.chdir(self.main)
+        run_command_with_process("npm run build:webpack")
+
+        logger.info("generate the `__init__.py` from template and verisons")
+        with open(self._concat(self.main, "init.template")) as fp:
+            t = string.Template(fp.read())
+
+        with open(self._concat(self.build_folder, "__init__.py"), "w") as fp:
+            fp.write(t.safe_substitute(versions))
+
+
+class Renderer(BuildProcess):
+    def __init__(self):
+        # dash-renderer's path is binding with the dash folder hierarchy
+        super(Renderer, self).__init__(
+            self._concat(
+                os.path.abspath(__file__),
+                os.pardir,
+                os.pardir,
+                os.pardir,
+                "dash-renderer",
+            ),
+            (
+                ("react", "umd", "react.production.min.js", None),
+                ("react", "umd", "react.development.js", None),
+                ("react-dom", "umd", "react-dom.production.min.js", None),
+                ("react-dom", "umd", "react-dom.development.js", None),
+                ("prop-types", None, "prop-types.min.js", None),
+                ("prop-types", None, "prop-types.js", None),
+            ),
+        )
+
+
+class DCC(BuildProcess):
+    def __init__(self, main=None):
+        main = (
+            os.path.abspath(os.path.expanduser(main)) if main else os.getcwd()
+        )
+        logger.info("start the build process at %s", main)
```

Comment:
unlike renderer, dcc folder cannot be computed from location of `__file__`,  user can either run it at root, or specify the path like this `dcc build --main=~/project/dash-core-components`

Reply:
I'm uneasy with putting dcc-specific code into the dash repo - can't this whole class live in the dcc repo? perhaps in the `scripts` dir? Then perhaps we wouldn't need this hack?

Reply:
I know the current code splits between two repos, but what about the part of the generation of Python & R components?  if we want to have a clean logical structure, we should probably 1: move all the relevant code into DCC 2. merge dcc into dash?

Reply:
Component generation is a generic process that applies to all component packages - and all of the package-specific pieces of that are controlled within the package it applies to. Initially that was done just by command-line args in the package scripts, now additional configuration is going into the `dash-info.yaml` file. There's nothing package-specific in `dash.development`. So if at all possible I'd like to continue that structure. That way other component packages can choose to use these facilities as well, including 3rd-party packages.

Reply:
sounds reasonable

OK,  I can move this class to dcc, also fix the issues @shammamah  reported in https://github.com/plotly/dash-core-components/issues/590

the dash is a dependency for dcc to run the generate-component and also import the base process class

Reply:
> the dash is a dependency for dcc to run the generate-component and also import the base process class

Right. It's currently an *unlisted* dependency, because we don't support dcc as an entry point and don't want a circular version conflict, you just get it when you install dash. Anyway the circularity is only there for devs, not for users. Is there a way to specify this kind of thing cleanly in `setup.py`?

Reply:
yes, the circular dependency was already present implicitly like this generated code in component

```from dash.development.base_component import Component, _explicitize_args```

but in dash and dash-renderer, actually, there is no such code to do direct import from DCC
we do that only with app writing. 

I feel DCC has more strong dependency on dash than the other direction.  Dash requires dcc mainly for app writing convenient.

Reply:
the only clear solution I see is merging dcc into dash :)